python/cog/server/clients.py

```diff
@@ -0,0 +1,258 @@
+        except ImportError:
+            pass
+        else:
+            return version("cog")
+        import pkg_resources
+
+        return pkg_resources.get_distribution("cog").version
+    except Exception:
+        return "unknown"
+
+
+_user_agent = f"cog-worker/{_get_version()} {httpx._client.USER_AGENT}"
+_response_interval = float(os.environ.get("COG_THROTTLE_RESPONSE_INTERVAL", 0.5))
+
+# HACK: signal that we should skip the start webhook when the response interval
+# is tuned below 100ms. This should help us get output sooner for models that
+# are latency sensitive.
+SKIP_START_EVENT = _response_interval < 0.1
+
+WebhookSenderType = Callable[[Any, WebhookEvent], Awaitable[None]]
+
+
+def webhook_headers() -> "dict[str, str]":
+    headers = {"user-agent": _user_agent}
+    auth_token = os.environ.get("WEBHOOK_AUTH_TOKEN")
+    if auth_token:
+        headers["authorization"] = "Bearer " + auth_token
+    return headers
+
+
+def httpx_webhook_client() -> httpx.AsyncClient:
+    return httpx.AsyncClient(headers=webhook_headers(), follow_redirects=True)
+
+
+def httpx_retry_client() -> httpx.AsyncClient:
+    # This session will retry requests up to 12 times, with exponential
+    # backoff. In total it'll try for up to roughly 320 seconds, providing
+    # resilience through temporary networking and availability issues.
+    transport = RetryTransport(
+        max_attempts=12,
+        backoff_factor=0.1,
+        retry_status_codes=[429, 500, 502, 503, 504],
+        retryable_methods=["POST"],
+    )
+    return httpx.AsyncClient(
+        headers=webhook_headers(), transport=transport, follow_redirects=True
+    )
+
+
+def httpx_file_client() -> httpx.AsyncClient:
+    # verify: Union[str, bool, ssl.SSLContext] = True
+    transport = RetryTransport(
+        max_attempts=3,
+        backoff_factor=0.1,
+        retry_status_codes=[408, 429, 500, 502, 503, 504],
+        retryable_methods=["PUT"],
+        verify=os.environ.get("CURL_CA_BUNDLE", True),
+    )
+    # set connect timeout to slightly more than a multiple of 3 to avoid
+    # aligning perfectly with TCP retransmission timer
+    # requests has no write timeout, keep that
+    # httpx default for pool is 5, use that
+    timeout = httpx.Timeout(connect=10, read=15, write=None, pool=5)
+    return httpx.AsyncClient(
+        transport=transport,
+        follow_redirects=True,
+        timeout=timeout,
+    )
+
+
+# there's a case for splitting this apart or inlining parts of it
+# I'm somewhat sympathetic to separating webhooks and files, but they both have
+# the same semantics of holding a client for the lifetime of runner
+# also, both are used by PredictionEventHandler
+
+
+class ClientManager:
+    def __init__(self) -> None:
+        self.webhook_client = httpx_webhook_client()
+        self.retry_webhook_client = httpx_retry_client()
+        self.file_client = httpx_file_client()
+        self.download_client = httpx.AsyncClient(follow_redirects=True)
```

Comment:
Extract this into a helper method like the others?

Also, initialize with `http2=True`

Reply:
:woman_shrugging: it's one line, all the other ones have more involved configuration